Formula generation support

shsms opened the pull request:
This PR adds formula generators for various microgrid metrics that traverse the graph to produce and return string formulas.

shsms committed:
Rename `graph::test_types` to `test_utils`

Also move the code for the module from `graph.rs` to
`graph/test_utils`.

This is to make room for a component graph builder for tests, that
will be added in the next commit.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `ComponentGraphBuilder` type in the `test_utils` module

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `Expr` type for representing formula expressions

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Support generating formulas with fallback expressions

Fallback expressions may be used when the primary expressions are
missing values.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add methods for iterating over the siblings of components

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Support searching the graph for components that match a predicate

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `consumer` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `grid` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `producer` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `battery` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `chp` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `PV` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add the `EV charger` formula generator

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Use `gh-action-cargo-test` in CI

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update documentation

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove `is_grid_meter` implementation

With the new consumer power formula, it is no longer necessary to
identify meters as grid meters.  So this method can go away.

If the SDK's traversal algorithms need it, they can use their own
temporary implementation, until they can migrate to the new formulas.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

ela-kotulska-frequenz commented on src/graph/test_utils.rs:
```diff
@@ -49,3 +54,162 @@ impl Edge for TestConnection {
+
+    /// Connects two components in the graph.
+    pub(super) fn connect(&mut self, from: ComponentHandle, to: ComponentHandle) -> &mut Self {
+        self.connections.push(TestConnection::new(from.0, to.0));
```
Minor, but if there is already interface to get component id then you could use it :)
`self.connections.push(TestConnection::new(from.component_id(), to.component_id()));`

shsms replied:
Done.

ela-kotulska-frequenz commented on src/graph/test_utils.rs:
```diff
@@ -49,3 +54,162 @@ impl Edge for TestConnection {
+        }
+        for battery in &batteries {
+            self.connect(inverter, *battery);
+        }
```
Minor but can be replaced by:
```
        for _ in 0..num_batteries {
            let battery = self.battery();
            self.connect(inverter, *battery);
            batteries.push(battery);
        }
```

shsms replied:
done.

ela-kotulska-frequenz commented on src/graph/test_utils.rs:
```diff
@@ -49,3 +54,162 @@ impl Edge for TestConnection {
+
+/// A builder for creating complex component graph configurations easily, for
+/// use in tests.
+pub(super) struct ComponentGraphBuilder {
+    components: Vec<TestComponent>,
+    connections: Vec<TestConnection>,
+    next_id: u64,
+}
```
Maybe this could be HashSet?
It will also secure us from bugs like having repeated component/connections in graph.

shsms replied:
No, we have test cases with duplicate components, so having it like this is useful.

ela-kotulska-frequenz commented on src/graph/formulas/expr.rs:
```diff
@@ -0,0 +1,379 @@
+    /// A subtraction of multiple expressions.
+    Sub { params: Vec<Expr> },
+
+    /// A coalesce function.
+    Coalesce { params: Vec<Expr> },
```
Please add description what Coalesce is :)

shsms replied:
I will write detailed docs on how to parse and use the formulas separately.

ela-kotulska-frequenz commented on src/graph/formulas/expr.rs:
```diff
@@ -0,0 +1,379 @@
+    }
+}
+
+#[derive(PartialEq)]
+enum BracketComponents {
+    First,
+    Rest,
+    All,
+    None,
+}
```
Could you add description?
It is not clear for me what these fields mean. And I have to guess reading the code

shsms replied:
Done.

ela-kotulska-frequenz commented on src/graph/formulas/expr.rs:
```diff
@@ -0,0 +1,379 @@
+    /// A numeric constant.
+    Number { value: f64 },
+
+    /// A reference to a component.
+    Component { component_id: u64 },
+
+    /// An addition of multiple expressions.
+    Add { params: Vec<Expr> },
+
+    /// A subtraction of multiple expressions.
+    Sub { params: Vec<Expr> },
+
+    /// A coalesce function.
+    Coalesce { params: Vec<Expr> },
```
I wonder if both `Coalesce` and `Component` should accept `default`.
default - the element which should be returned if  component send no data

shsms replied:
This can be `COALESCE(..., 0.0)`, which is there for some components already.  But I'll do that in the fallback method, so it is available for all.

shsms replied:
This is in progress, will add a new commit.

ela-kotulska-frequenz commented on src/graph/formulas/fallback.rs:
```diff
@@ -0,0 +1,233 @@
+        Ok(exprs)
+    }
+
+    fn meter_fallback(&self, component_id: u64) -> Result<Option<Expr>, Error> {
+        let component = self.graph.component(component_id)?;
+        if component.is_meter()
+            && self.graph.has_successors(component_id)?
+            && !self.graph.has_meter_successors(component_id)?
```
Could you add documentation what this function should do?
It is very nested and it took me long time to read it.

shsms replied:
I added some basic docs but also refactored all the methods in this file, so hopefully, they're easier to read.

llucax reviewed (commented):
I gave this a quick look and looks really nice, with all the automatic fallaback/`COALESCE`. As a no-rust, it feels a bit the docs are not very verbose though. Can we maybe generate the docs as GitHub pages? Or will they be automatically generated and published when uploading to crates.io?

It looks to me that maybe more in depth docs about how do the formula builders work (what kind of components are considered, which are used as fallback of which, etc.) might be helpful for users to know.

llucax commented on README.md:
```diff
@@ -6,7 +6,7 @@ connections between them as a Directed Acyclic Graph (DAG).
 Current features:
   - validating the graph
   - traversing the graph
+  - Formula generation
```
nitpick: consistent upper or lower case :D

shsms replied:
Fixed.

ela-kotulska-frequenz commented on src/graph/formulas/fallback.rs:
```diff
@@ -0,0 +1,233 @@
+            && self.graph.has_successors(component_id)?
+            && !self.graph.has_meter_successors(component_id)?
+        {
+            if self
+                .graph
+                .successors(component_id)?
+                .all(|x| x.is_supported())
+            {
+                let mut exprs = vec![
+                    Expr::components(
+                        self.graph
+                            .successors(component_id)?
+                            .map(|x| x.component_id()),
+                    )
+                    .into_iter()
+                    .reduce(|a, b| a + b)
+                    .ok_or(Error::internal(
+                        "Can't find successors of components with successors.",
+                    ))?,
+                    Expr::component(component_id),
+                ];
```
Either I don't understand this fragment or it is over complicated.
Would something like this work? It is just an idea
```
let successors = self.graph.successors(component_id)?;
if successors.len() == 0 || successors.all(|x| x.is_supported()) == false:
      return Ok(Some(Expr::component(component_id)));

let fallback_expr = successors.map(|node|  Expr::component(node.component_id)).reduce(|a, b| a + b))? 
exprs = match self.prefer_meters:
   true => vec![Expr::component(component_id), fallback_expr]
   false => vec![ fallback_expr, Expr::component(component_id)]
return  Ok(Some(Expr::coalesce(exprs)
```

shsms replied:
True, it was possibly too nested. I've simplified this as suggested.

ela-kotulska-frequenz replied:
One more comment to this below.

ela-kotulska-frequenz reviewed (commented):
I survived until commit: `Support generating formulas with fallback expressions ` :)
Will check rest later today or tomorrow.

shsms commented:
> As a no-rust, it feels a bit the docs are not very verbose though. Can we maybe generate the docs as GitHub pages? Or will they be automatically generated and published when uploading to crates.io?

They will be generated from our inline doc comments and become available at docs.rs when we publish to crates.io.

> 
> It looks to me that maybe more in depth docs about how do the formula builders work (what kind of components are considered, which are used as fallback of which, etc.) might be helpful for users to know.


More documentation is necessary, especially definitions of formula components, instructions for using the formulas, and explanations for the algorithms. But I'll get to those after the formulas and algorithms have stabilized; otherwise, I'll spend all my time rewriting the docs.

ela-kotulska-frequenz commented on src/graph/formulas/fallback.rs:
```diff
@@ -0,0 +1,238 @@
+            return Ok(None);
+        }
+
+        if !self.graph.has_successors(component_id)?
+            || !self
+                .graph
+                .successors(component_id)?
+                .all(|x| x.is_supported())
+        {
+            return Ok(Some(Expr::component(component_id)));
+        }
+
+        let successor_expr = self
+            .graph
+            .successors(component_id)?
+            .map(Expr::from)
+            .reduce(|a, b| a + b)
+            .ok_or(Error::internal(
+                "Can't find successors of components with successors.",
```
Why do you call `self.graph.has_successors(component_id)` 3 times?
My example called it once and it makes code 7 lines shorter and more readable :)

shsms replied:
`graph.successors` returns an iterator, not a set.  We can't do `len() == 0`, etc., on it, but the `has_successors` method is still O(1) because it only checks if there's at least one successor.

Because it doesn't allocate and allows us to iterate on nodes stored in the graph, this is better than making a local copy.

ela-kotulska-frequenz replied:
Thanks a lot for explanation!

ela-kotulska-frequenz commented on src/graph/formulas/fallback.rs:
```diff
@@ -0,0 +1,238 @@
+        // component ids, the predecessors can't be used as fallback.
+        let siblings = self
+            .graph
+            .siblings_from_predecessors(component_id)?
```
This and next commit are in wrong order. :)
Could you please switch them? Otherwise it is hard to read.

shsms replied:
Fixed.

ela-kotulska-frequenz commented on src/graph/retrieval.rs:
```diff
@@ -105,6 +106,41 @@ where
+    /// from the component with the given `component_id`.
+    ///
+    /// If `follow_after_match` is `true`, the search continues deeper beyond
+    /// the matching components.
+    pub(crate) fn find_all(
+        &self,
+        from: u64,
+        mut pred: impl FnMut(&N) -> bool,
+        follow_after_match: bool,
```
Minor: I don't understand why it is called `find_all`? 
Because It doesn't finds all nodes. It finds only nodes in direction outgoing. :) 

Maybe `find_with_pred` would be more descriptive?

shsms replied:
Initially, it also had a direction parameter, but I realized we didn't need it.  I will think of an option.

shsms replied:
I've added a direction parameter as a filter instead.

ela-kotulska-frequenz reviewed (commented):
Checked until  `Add the consumer formula generator`

shsms committed:
Add additional per-component fallback to 0.0

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Remove `Node::is_supported` trait method

All components now have a fallback to 0.0, so this check is no longer
necessary.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

ela-kotulska-frequenz commented on src/graph/formulas/generators/consumer.rs:
```diff
@@ -0,0 +1,435 @@
+                // difference of "mixed" meter from its successors.
+                "COALESCE(MAX(0.0, #11 - #8 - #9 - #10), 0.0)"
+            )
+        );
```
I am wondering why we are using meters are primary components and inverters as fallback components?
Maybe it would be easier to use inverters are primary and meters as fallback components?

In current version  if meters are primary components and working: we need data from both meters and its successors to get accurate output. And if any meter is not working, we use only inverter data and it gives accurate output, too => less data and the same accuracy. 

But if inverters were primary components and working: we would not need meter data until any inverter stop working.

shsms replied:
I agree that the formula can be simplified as you suggest, but I'd like to make an issue for now because the current one is still correct, and the simplified formula would need a rewrite.

shsms replied:
https://github.com/frequenz-floss/frequenz-microgrid-component-graph-rs/issues/4

ela-kotulska-frequenz commented on src/graph/formulas/generators/consumer.rs:
```diff
@@ -0,0 +1,435 @@
+            }
+        }
+
+        let other_grid_successors = self
+            .graph
+            .successors(self.graph.root_id)?
+            .filter(|s| !s.is_meter() && !s.is_battery_inverter())
+            .map(|s| self.component_consumption(s.component_id()))
+            .reduce(|a, b| Ok(a? + b?));
```
Why battery_inverters are excluded here?

shsms replied:
Battery consumption is not site consumption because it can be discharged from the battery.  The formula is only for site consumption, not intermediate storage.

shsms committed:
Add a direction parameter to the `find_all` method

Earlier the direction was hard coded, which made it unclear that the
method was only finding outgoing nodes.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>
